File: Structy/Recursion_Beginner/ReverseStringRecursive.java
class Source {
  public static String reverseString(String s) {
    String result;
    int length = s.length();
    if(s.length() == 0){
      return "";    
    }
    result = reverseString(s.substring(1)) + s.charAt(0);
    return result;
  }

  public static void run() {
    // this function behaves as `main()` for the 'run' command
    // you may sandbox in this function , but should not remove it
  }
}

/*
  Call stack is required to store
  Time complexity O(n^2)
  Space Complexity O(n^2)

*/

File: Structy/Python/Recursion/ReverseString_Recursive.py
def reverse_string(s):
  if len(s) == 0:
    return ""
  return reverse_string(s[1:]) + s[0]

# n is the length of the string
#Time Complexity: O(n^2)
#Space Complexity: O(n^2)
#Reason: We are using a recursive function to reverse the string.